evm/src/witness/memory.rs

```diff
@@ -0,0 +1,83 @@
+use std::collections::HashMap;
+
+use ethereum_types::U256;
+
+use crate::cpu::membus::{NUM_CHANNELS, NUM_GP_CHANNELS};
+
+pub enum MemoryChannel {
+    Code,
+    GeneralPurpose(usize),
+}
+
+use MemoryChannel::{Code, GeneralPurpose};
+
+impl MemoryChannel {
+    pub fn index(&self) -> usize {
+        match *self {
+            Code => 0,
+            GeneralPurpose(n) => {
+                assert!(n < NUM_GP_CHANNELS);
+                n + 1
+            }
+        }
+    }
+}
+
+pub type MemoryAddress = (u32, u32, u32);
+
+pub enum MemoryOpKind {
+    Read,
+    Write(U256),
+}
+
+pub struct MemoryOp {   
+    pub timestamp: u64,
+    pub address: MemoryAddress,
+    pub op: MemoryOpKind,
+}
+
+impl MemoryOp {
+    pub fn new(
+        channel: MemoryChannel,
+        clock: usize,
+        address: MemoryAddress,
+        op: MemoryOpKind,
+    ) -> Self {
+        let timestamp = (clock * NUM_CHANNELS + channel.index()) as u64;
+        MemoryOp { timestamp, address, op }
+    }
+}
+
+
+#[derive(Clone)]
+pub struct MemoryState {
+    contents: HashMap<MemoryAddress, U256>,
```

Comment:
I think a dense representation of memory (like the old `MemorySegmentState`) would probably be more efficient. I wouldn't expect much sparseness based on the cost model, since it charges based on the highest address used.